### ISSUE TITLE
Allow editing of text after coding of text

### ISSUE THREAD

**Issue by jdittrich** [openqda/openqda#61]
* **Story**: I want to be able to correct spelling or transcription mistakes in my texts
* **Current behavior:** Files can not be unlocked after coding, they can only be deleted
* **Expected behavior:** I can unlock texts and edit them (similar to texts in e.g. AtlasTI)
* **Reproduction:** 
  * Load a text
  * Apply a code
  * Try to edit the text

**Comment by alessandrobelli** [comment]
This is definitely something we want to add, but it adds layers of complexity.
For example, the codes are identified by the start and end position in a text, and their length. While the length change is easy, if the text is altered in a considerable way, this can easily become a problem.

TLDR; It's on our schedule, but it won't be available anytime soon.

Unfortunately for now, the workflow has to include a strong review of the text before coding starts...

**Comment by jdittrich** [comment]
>  While the length change is easy, if the text is altered in a considerable way

Yes – the method I found to do this (and there might be more efficient ones):

* Have the text changes in the form as objects with the properties _length:int>_ and _operation-type_ (which are: "retain", "delete", "add") – this is what many modern editors produce (I think I used quill 1)
* Represent the _annotate ranges_ of of text as _boundaries_, i.e. an objects with the properties _boundary position index:int_, _boundary type_ (which are: "start", "end") and _annotation id_  (that usually needs a temporary conversion)
* For each change go though all boundaries and see if they need to be moved. A boundary is moved, if its _position index_ is > than the position-index of the change. Deletions lower all boundary indices, additions raise them. 
* …what I did not implement is cleaning up collapsed boundaries i.e. ones where their containing text has been entirely deleted.
* convert back into the native format of representing annotation ranges.

Which does not mean this is easy to do, just some example of how it can be approached.

**Comment by jankapunkt** [comment]
@kadewe the aforementioned comment by  @jdittrich is an example of why we now use quill to manage the state of selections as controlled state as opposed to manual add/insert/update/remove elements.

The clear advantage is that quill supports concurrent editing using a delta format, which will soon allow us to make selections of text, then edit the text and have the selections automatically update.

However, there is one more level of complexity that @jdittrich missed out here, that is - we have to additionally compute overlaps of selections, because quill and any other editor will not automatically create an overlap, because it's actually their job to NOT overlap but override. Think about how google docs would work if the algorithm (operational transformation) would actually support overlapping editing, instead of overriding... This is why "merge" of concurrent edits is simply a controlled way of deciding which edit is the correct last state of truth overriding prior edits.

Summarizing this:

- from version 1.0 we use quill to control the state of the text and the selections
- with quill we can support multiple plugins for visualizing selections, the default (and currently only supported one) is highlighted background
- overlapping of selections has to be computed manually and has to be added until 1.0
- additional deltas from post-selection editing of the source should be implemented but not for 1.0
- performance might be affected on large documents with lots of selections by multiple users, this is to be researched aftr 1.0 release in order to find performance optimizations

**Comment by jankapunkt** [comment]
@jdittrich overlapping selections might not be part of your usecase but since collaborative coding will be a crucial part of the upcoming releases we have to cover visualization of overlapping codes. Yes, the overlappings are only part of the view-model but they have to be actively computed each time a selection gets added/removed/moved, because the selection itself does not know which other selections are crossed.

A demo would be very useful, can you link it here?

**Comment by jdittrich** [comment]
> overlapping selections might not be part of your usecase

They are! (I just derived them, rather than represent them directly)

> Yes, the overlappings are only part of the view-model but they have to be actively computed each time a selection gets added/removed/moved, because the selection itself does not know which other selections are crossed.

Yes.

> A demo would be very useful, can you link it here?

I could send code, I think, I do not have an online demo for it.

**Comment by jankapunkt** [comment]
@jdittrich I could create a demo repository and invite you and we build this together until it's a proof of concept. We could then merge it into openqda, making you our first external collaborator :rocket:

**Comment by jankapunkt** [comment]
@jdittrich here is the repo: https://github.com/openqda/editor-dev let me know if I can invite you :-)

**Comment by jdittrich** [comment]
thanks – yes, please invite me (the account that I use here is right)

**Comment by jankapunkt** [comment]
@jdittrich done, let's continue discussion there and come back here later
